coins/src/utils/coins3/produce.ts

```diff
@@ -15,102 +18,170 @@ export type Dynamo = {
+    if (body.length) {
+      await esClient.bulk({ body });
+    }
+  } catch (err) {
+    console.error("Metadata fallback (bulk to ES) failed:", err);
+  }
+}
+
+async function produceTopics(items: Dynamo[], topic: Topic, producer: Producer) {
+  const messages: string[] = [];
+
+  items.map((item) => {
+    const { price, SK } = item;
+
+    if (topic !== "coins-timeseries" && SK !== 0) return;
+    if (topic !== "coins-metadata" && !price) return;
+    if (topic === "coins-metadata") return;
+
+    const messageObject = convertToMessage(item, topic);
+    if (!messageObject) return;
+
+    validate(messageObject, topic);
+
+    const message = JSON.stringify(messageObject);
+    if (!messages.includes(message)) {
+      messages.push(message);
+    }
+  });
+
+  if (process.env.KAFKA_UP === "true") {
+    try {
+      await producer.send({
+        topic: `${topic}`,
+        messages: messages.map((value) => ({ value })),
+      });
+      return;
+    } catch (error) {
+      console.error(`Kafka failed for topic "${topic}". Fallback to ES/Redis.`, error);
+    }
   }
 
-  if (pk.startsWith(assetPrefix)) pk = pk.substring(assetPrefix.length);
-  const chain = pk.split(":")[0].toLowerCase();
-  const address = pk.substring(pk.split(":")[0].length + 1).toLowerCase();
-  return { chain, address, pid: `${chain}:${address}` };
+  if (topic === "coins-current") {
+    try {
+      const redisClient = getRedis();
+      const pipeline = redisClient.pipeline();
+      messages.forEach((msg) => {
+        const parsed = JSON.parse(msg);
+        pipeline.set(`price_${parsed.pid}`, msg, "EX", 24 * 3600);
+      });
+      await pipeline.exec();
+    } catch (redisError) {
+      console.error("Redis fallback failed for coins-current:", redisError);
+    }
+  } else if (topic === "coins-timeseries") {
+    try {
+      const body: any[] = [];
+      messages.forEach((msg) => {
+        const parsed = JSON.parse(msg);
+        const date = new Date(parsed.ts * 1000);
+        const year = date.getFullYear();
+        const month = String(date.getMonth() + 1).padStart(2, "0");
+        const index = `coins-timeseries-${year}-${month}`;
+
+        body.push({ index: { _index: index, _id: `${parsed.pid}_${parsed.ts}` } });
+        body.push(parsed);
+      });
+      if (body.length) {
+        await esClient.bulk({ body });
+      }
+    } catch (esError) {
+      console.error("ES fallback failed for coins-timeseries:", esError);
+    }
+  }
 }
 
 export default async function produce(
   items: Dynamo[],
-  topics: Topic[] = allTopics,
+  topics: Topic[] = allTopics
 ) {
   try {
     if (!items.length) return;
-    const invalidTopic = topics.find((t: any) => {
-      !allTopics.includes(t);
-    });
+    const invalidTopic = topics.find((t: any) => !allTopics.includes(t));
     if (invalidTopic) throw new Error(`invalid topic: ${invalidTopic}`);
+
     const producer: Producer = await getProducer();
+    if (topics.includes("coins-metadata")) {
```

Comment:
we want to check metadata record for all topics right? if so, why add this check?